src/test/resources/test-scripts/scripted_object.bsh

```diff
@@ -0,0 +1,39 @@
+#!/bin/java bsh.Interpreter
+
+source("TestHarness.bsh");
+
+/*
+ * Test that variable lookups in scripted objects work correctly
+ * Related to #659
+ */
+
+test1(String c) {
+   return this;
+}
+
+container() {
+   ylds = new ArrayList();
+   return this;
+}
+
+Object c = null;
+{
+   {
+      c = container();
+   }
+   c.ylds.add(test1("hello"));
+}
```

Comment:
This I can agree with but if I write 
```
Object c = null;
{
    {
        Object c = container();
    }
    assert(c == null);
}
```

If you redeclare an existing parent variable it must hide the parent. It is how local variable access and inheritance works in JAVA.

Take for example:
```java
jshell> static int getInt() {
...>     int[] i = {1};
...>     class A {
...>         A() {
...>            i[0] = 2;
...>         }
...>     }
...>     new A();
...>     return i[0];
...> }
|  replaced method getInt()

jshell> getInt();
$21 ==> 2
```
Local variable changed in parent namespace.

And then I decide to declare my own variable in the block namespace.
```java
jshell> static int getInt() {
...>     int[] i = {1};
...>     class A {
...>         A() {
...>            int[] i = new int[1];
...>            i[0] = 2;
...>         }
...>     }
...>     new A();
...>     return i[0];
...> }
|  replaced method getInt()

jshell> getInt();
$23 ==> 1
```

Then it does not change the parent which remains 1, but the value in the constructor block is 2 and not 1 that would be wrong, don't you agree?.

Reply:
Inheritance:
```java
jshell> class A { int i = 1;}
|  created class A

jshell> class B extends A {
...>      B() {
...>         i = 2;
...>      }
...>      int getInt() {
...>         return i;
...>      }
...> }
|  replaced class B

jshell> new B().getInt();
$29 ==> 2

jshell> ((A)new B()).i
$30 ==> 2
```
Parent namespace declared the variable so it changed.

But if the child namespace declares the variable:

```java
jshell> class A { int i = 1;}
|  created class A

jshell> class B extends A {
...>      int i = 0;
...>      B() {
...>         i = 2;
...>      }
...>      int getInt() {
...>         return i;
...>      }
...> }
|  replaced class B

jshell> new B().getInt();
$32 ==> 2

jshell> ((A)new B()).i
$33 ==> 1
```

The child namespace value changed but the parent variable was hidden and untouched.

Reply:
> If you redeclare an existing parent variable it must hide the parent.

no disagreement with your first example. 

> Then it does not change the parent which remains 1

No disagreement with the second example

> Parent namespace declared the variable so it changed.

No disagreement with the third example

> The child namespace value changed but the parent variable was hidden and untouched.

No disagreement with the fourth example.

I am not sure what point you are trying to make here.  I am not disagreeing with how Java works.  The issue that I was raising was in regards to the new semantics introduced in to beanshell, and how they differ from java and previous beanshell.